### PR TITLE
[v2.7] Add logic and CI checks for image source code origins

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
This PR accompanies the upcoming trusted registry and associated work targeted for 2.7.0. 

This PR introduces a new build artifact for Rancher, named `rancher-origins.txt`. This artifact contains a mapping between all images used in Rancher and their origin repository. This allows users of Rancher to look at the source code of any image being used by an out-of-the box Rancher installation. All images which are produced in `rancher-images.txt` & `rancher-windows-images.txt` are expected to have a URL pointing to their source code repository. 

This PR also introduces an additional CI check to ensure that all images have a URL associated with them, failing builds if images are unaccounted for. If all images are accounted for then the new artifact will be generated and have a similar structure to `rancher-sources.txt` - a space delimited list of images and their source code repository URL's. 

A tool has also been written which will attempt to automatically resolve source code repositories for some images, however it is primarily written to resolve Rancher owned images which share a name with their base repository. A warning will also be printed out which denotes images that are no longer used but are still present within `pkg/image/origins.go`. This tool can be run with the command `dapper check-origins`, and is helpful for addressing any CI errors that may occur due to this new process. 

This mapping of images to source code URL's is maintained in `pkg/image/origins.go`, in the variable titled `OriginMap`. This map does not take into account image versions, and only needs to be updated when net-new images are added to Rancher. When images are no longer used by Rancher they should be removed from `pkg/image/origins.go`, however leaving them in will not cause a build failure (though these unused images _will_ be included in the release artifact). The command `dapper check-origins` will print a warning if it finds any images which are no longer used by rancher.

I've done my best to identify the source code repository for each of the images within the currently released `rancher-images.txt` & `rancher-windows-images.txt`, and have updated the map within `/pkg/image/origins.go`. However, this map should be double-checked to ensure that I have found the correct repository for each image, and if not comment on this PR with the correct URL so that I may update it.

### CODE REVIEW

File: pkg/image/origins.go
Comment (jakefhyde) [inline pkg/image/origins.go:242]: QQ: how long does this whole script take to run?

Comment (HarrisonWAffel) [reply pkg/image/origins.go:242]: If you have the image needed for dapper already cached then it takes about 1 minute, most of that time is spent cloning the charts repo and unpacking it so we can get at the images. If you don't have it cached it might take 2-3 minutes, but again that's largely due to dapper setting up and cloning the repositories which can't really be avoided.

File: pkg/image/origins.go
Comment (jakefhyde) [inline pkg/image/origins.go:295]: I think you can replace this with `map[string]struct{}`, which is idiomatic for set containers.

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/image/origins/main.go
Comment (jiaqiluo) [inline pkg/image/origins/main.go:274]: The same function is already defined in `pkg/image/export/main.go`. 
How about creating a new file `pkg/image/utilities.go` for all common helper functions?

File: pkg/image/origins/main.go
Comment (jiaqiluo) [inline pkg/image/origins/main.go:37]: Most of this function is the same logic as the function `run` in `pkg/image/export/main.go` 
How about extracting the common parts to make a helper function?

File: scripts/check-origins
Comment (jiaqiluo) [inline scripts/check-origins:1]: This file is almost identical to the file `scripts/package`,
how about moving the shared parts into another script file? 

Another reason for the suggestion is that today we need to change the default value of `SYSTEM_CHART_DEFAULT_BRANCH` between `dev-v2.x` and `release-v2.x` in the file `scripts/package` before/after releasing rancher.  If we add this file, it means one more place for us to change the same variable.

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: scripts/check-origins
Comment (HarrisonWAffel) [reply scripts/check-origins:1]: I'll move the common code into a shared script, though the `check-origins` script is expected to only be used in a developer workflow, it doesn't contribute to a release or produce any artifacts on its own so I wasn't anticipating needing to update it.

Comment (HarrisonWAffel) [reply scripts/check-origins:1]: A new file, `package-env`, was created to handle the common environment variables between the two scripts

File: pkg/image/origins/main.go
Comment (HarrisonWAffel) [reply pkg/image/origins/main.go:31]: `utilities.go` now handles the processing of the images and image sources within `GatherTargetImagesAndSources`, which is used by both `img/export/main.go` and `img/origins/main.go`.

File: pkg/image/origins.go
Comment (HarrisonWAffel) [reply pkg/image/origins.go]: I've changed this in two placed, both `UniqueTargetImages` and `CheckForImagesNoLongerBeingUsed`.

File: pkg/image/origins/main.go
Comment (HarrisonWAffel) [reply pkg/image/origins/main.go]: Created `utilities.go` to handle shared logic 👍

File: pkg/image/origins.go
Comment (HarrisonWAffel) [reply pkg/image/origins.go]: I've update this for both places where we identify unique images

File: pkg/image/utilities/utilities.go
Comment (jiaqiluo) [inline pkg/image/utilities/utilities.go:63]: nit: `ioutil.ReadFile` is deprecated, this function simply calls `os.ReadFile`.

File: pkg/image/export/main.go
Comment (jiaqiluo) [inline pkg/image/export/main.go:47]: nit 
```suggestion
		if err = utilities.ImagesAndSourcesText(arch, imageLists.imagesAndSources); err != nil {
```

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/image/export/main.go
Comment (HarrisonWAffel) [reply pkg/image/export/main.go]: good catch

**Review by jiaqiluo — APPROVED** [review]
LGTM.

@HarrisonWAffel , there is one more follow-up after the PR is merged:
The releasing checklist in Confluence needs to be updated because this PR moves variables from `scripts/package` to `scripts/package-env`. 

You might need to reach out to Jono for all affected files.

File: pkg/image/origins/main.go
Comment (jakefhyde) [inline pkg/image/origins/main.go:160]:
```suggestion
		case strings.Contains(repo, "hardened"):
			// hardened images conversion, best effort
			repo = strings.ReplaceAll(repo, "hardened-", "image-build-")
			fallthrough
```
I'm not saying that you have to, just that it's possible.

**Review by jakefhyde — APPROVED** [review]

File: pkg/image/origins/main.go
Comment (HarrisonWAffel) [reply pkg/image/origins/main.go:160]: Yeah, I had thought about that too its just that some of the hardened images have slightly different names than their base repo's, for instance the `hardened-dns-node-cache` comes from `https://github.com/rancher/image-build-dns-nodecache`, which doesn't have a `-` separating `node` and `cache`.

I'll do a double check of the hardened images tomorrow to make sure I have them all correct, the list still isn't completely finalized

**Review by jakefhyde — DISMISSED** [review]

**Review by superseb — COMMENTED** [review]
Was it a design choice to have all mappings in the code? There is code to derive the source repository from the repository name but I still see the mappings that match that in the code mappings while these should be accounted for. I believe it is more confusing to have entries in the mappings that can be derived as we call out we have logic for those cases. 

And some images might have `org.opencontainers.image.url` pointing to the source repository, is there a specific reason not to use this as a source for finding out the source? I guess this builds on the previous answer, do we want a full list with origins or auto-detect as much as possible and have the list of exceptions that can't be derived.

File: pkg/image/origins.go
Comment (jakefhyde) [inline pkg/image/origins.go:163]: Is it feasible to add a check to `convertImagesToRepoUrls` to check for this the same way we do for `hardened`? Seems mirrored images could potentially check `rancher/<REPO>`, then `<ORG>/<REPO>` before setting as unknown.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
The Team1 managed charts look good. 

Below are some nits because I don't know how short our timeline is for this PR:
- We should have comments on all exported functions.  https://go.dev/doc/comment
- This still needs unit tests.
- error wrapping on errors returned by external packages.

**Comment by HarrisonWAffel** [comment]
@superseb the reason I've included them all in the map is that I'm trying to avoid lengthening the time it takes for CI to run as much as possible. Currently the means in which we automatically resolve image origins is via synchronous `GET` requests, and as more images are added that will add time to each build. I'm open to ideas on how to refine this, I just don't want to extend the time it takes to complete CI due to this artifact.

Also, it may be difficult to find the actual source code of a particular image if we simply state its source as another image, which is why i opted for URL's when possible. ideally, we want to cover all images specified in `rancher-images.txt` and `rancher-windows-images.txt` so that when customers pull these images into their own private repositories they can verify what is being executed.

File: pkg/image/origins.go
Comment (HarrisonWAffel) [reply pkg/image/origins.go]: Good idea, added in latest commit

File: pkg/image/utilities/utilities.go
Comment (HarrisonWAffel) [reply pkg/image/utilities/utilities.go]: I've added the aggregate type `ImageTargetsAndSources` to handle those slices

**Comment by KevinJoiner** [comment]
When we are doing error wrapping we should use `fmt.Errorf("%w")`
We do not want to use the package `github.com/pkg/errors` it was made obsolete by go 1.13 when they added error wrapping to the standard lib. And the package has since been archived.

File: pkg/image/origins/main.go
Comment (jakefhyde) [inline pkg/image/origins/main.go:155]: `https://www.github.com` just 301s to `https://github.com`. I don't think it will be a measurable time save, but might be worth trying.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/image/origins.go
Comment (HarrisonWAffel) [reply pkg/image/origins.go:235]: This was done to ensure the log message matches the format of existing log messages, I noticed it didnt when addressing the above comments. 

```
INFO[0000] generating k3s image list...                 
INFO[0006] finished generating k3s image list...        
INFO[0006] generating rke2 image list...                
INFO[0019] finished generating rke2 image list...       
INFO[0019] building rancher-image-origins.txt           // doesn't match the below messages 
2022/10/07 21:57:58 Creating rancher-images.txt
2022/10/07 21:57:58 Creating rancher-images-sources.txt
2022/10/07 21:57:58 Creating rancher-mirror-to-rancher-org.sh
2022/10/07 21:57:58 Creating rancher-save-images.sh
2022/10/07 21:57:58 Creating rancher-load-images.sh
2022/10/07 21:57:58 Creating rancher-windows-images.txt
2022/10/07 21:57:58 Creating rancher-windows-images-sources.txt
2022/10/07 21:57:58 Creating rancher-mirror-to-rancher-org.ps1
2022/10/07 21:57:58 Creating rancher-save-images.ps1
2022/10/07 21:57:58 Creating rancher-load-images.ps1
```

Is now 

```
INFO[0000] generating k3s image list...                 
INFO[0003] finished generating k3s image list...        
INFO[0003] generating rke2 image list...                
INFO[0012] finished generating rke2 image list...       
2022/10/11 15:23:51 building rancher-image-origins.txt
2022/10/11 15:23:51 Creating rancher-images.txt
2022/10/11 15:23:51 Creating rancher-images-sources.txt
2022/10/11 15:23:51 Creating rancher-mirror-to-rancher-org.sh
2022/10/11 15:23:51 Creating rancher-save-images.sh
2022/10/11 15:23:51 Creating rancher-load-images.sh
2022/10/11 15:23:51 Creating rancher-windows-images.txt
2022/10/11 15:23:51 Creating rancher-windows-images-sources.txt
2022/10/11 15:23:51 Creating rancher-mirror-to-rancher-org.ps1
2022/10/11 15:23:51 Creating rancher-save-images.ps1
2022/10/11 15:23:51 Creating rancher-load-images.ps1

```

File: pkg/image/origins.go
Comment (HarrisonWAffel) [reply pkg/image/origins.go]: I added that logic in to the latest commits 👍

**Review by KevinJoiner — APPROVED** [review]

**Review by jakefhyde — APPROVED** [review]

**Review by jiaqiluo — APPROVED** [review]

**Comment by HarrisonWAffel** [comment]
Could someone from @rancher/rancher2-helm-yes-dev please take a look at this PR before EOD PST to check that the URL's I've provided are correct? As tomorrow is code freeze and I've gotten approvals from team 1 and team 2 I'll be merging this PR EOD PST, but would appreciate a +1 from team 3 so that they're aware of the new CI check.

File: pkg/image/origins.go
Comment (aiyengar2) [inline pkg/image/origins.go:201]: This comes from github.com/rancher/PushProx

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
PushProx images have the wrong source, otherwise LGTM

**Review by superseb — APPROVED** [review]

**Review by KevinJoiner — APPROVED** [review]

**Pull request merged**
